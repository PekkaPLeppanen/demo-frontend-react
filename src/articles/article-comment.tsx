import { type IArticleComment } from 'resources/article-provider'
import 'articles/article-comment.scss'

export interface ArticleCommentProps {
  comment: IArticleComment
}

export function ArticleComment({comment}: ArticleCommentProps): JSX.Element {
  return (
    <div className="article-comment">
      <div className="article-comment__email">{comment.email}</div>
      <div className="article-comment__body">{comment.body}</div>
    </div>
  )
}
