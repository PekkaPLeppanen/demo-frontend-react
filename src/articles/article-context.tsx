import { type IArticle } from 'resources/article-provider'
import { createContext, PropsWithChildren, useContext, useState } from 'react'

export type IArticleContext = {
  selectedArticle: IArticle | null;
  selectArticle: (article: IArticle) => void;
};

const ArticleContext = createContext<IArticleContext | null>(null)

/**
 * Creates a context provider component, where all the children have access to the ArticleContext
 */
export function ArticleProvider(props: PropsWithChildren): JSX.Element {

  const [article, selectArticle] = useState<IArticle | null>(null)

  return (
    <ArticleContext.Provider value={{selectedArticle: article, selectArticle}}>
      {props.children}
    </ArticleContext.Provider>
  )
}

/**
 * Usage: const {article, selectArticle} = useArticle();
 */
export function useArticle(): IArticleContext {
  return useContext(ArticleContext) as IArticleContext
}
