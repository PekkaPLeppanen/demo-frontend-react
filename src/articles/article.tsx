import { getArticleComments, IArticleComment } from 'resources/article-provider'
import { ArticleComment } from 'articles/article-comment'
import { useEffect, useState } from 'react'
import 'articles/article.scss'
import { useArticle } from './article-context'

export function Article() {

  const [comments, setComments] = useState<IArticleComment[]>()
  const {selectedArticle} = useArticle()

  useEffect(() => {
    if (!selectedArticle) {
      return
    }
    getArticleComments(selectedArticle?.id)
      .then(response => {
        setComments(response ?? [])
      })
  }, [selectedArticle])

  return (
    <div className="article">
      <h2 className="article__title">{selectedArticle?.title}</h2>
      <p className="article__body">{selectedArticle?.body}</p>
      {
        comments
          ? comments.map(comment => <ArticleComment comment={comment} key={comment.id}/>)
          : <p>Loading comments...</p>
      }
    </div>
  )

}
