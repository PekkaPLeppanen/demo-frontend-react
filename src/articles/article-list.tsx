import { getAllArticles, IArticle } from 'resources/article-provider'
import 'articles/article-list.scss'
import React, { useEffect, useState } from 'react'
import { Article } from 'articles/article'
import { useArticle } from 'articles/article-context'

export function ArticleList(): JSX.Element {

  const [articles, setArticles] = useState<IArticle[]>([])
  const {selectedArticle, selectArticle} = useArticle()

  useEffect(() => {
    getAllArticles()
      .then(articles => {
        setArticles(articles)
      })
  }, [])

  return (
    <>
      <h1>Latest topics</h1>
      {selectedArticle && <Article />}
      {articles.length === 0
        ? <p>Loading...</p>
        : <ul className="article-list">
          {articles.map((article) =>
            (
              <li className="article-list__item" key={article.id} onClick={() => selectArticle(article)}>
                {article.title}
              </li>
            ))}
        </ul>
      }
    </>
  )
}
