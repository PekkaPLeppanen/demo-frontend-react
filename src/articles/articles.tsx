import { getAllArticles, IArticle } from 'resources/article-provider'
import 'articles/article-list.scss'
import React, { useEffect, useState } from 'react'

export function Articles(): JSX.Element {

  const [articles, setArticles] = useState<IArticle[]>([])

  useEffect(() => {
    getAllArticles()
      .then(articles => {
        setArticles(articles)
      })
  }, [])

  return (
    <>
      <h1>Latest topics</h1>
      {articles.length === 0
        ? <p>Loading...</p>
        : <ul className="article-list">
          {articles.map((article) =>
            (
              <li className="article-list__item" key={article.id}>
                {article.title}
              </li>
            ))}
        </ul>
      }
    </>
  )
}
