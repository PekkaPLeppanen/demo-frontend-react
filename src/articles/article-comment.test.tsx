import { ArticleComment } from 'articles/article-comment'
import { articleCommentsResponseMock } from 'resources/test-data/articles'
import { render, screen } from '@testing-library/react'

describe('<ArticleComment />', () => {
  test('render', async () => {
    const articleComment = articleCommentsResponseMock[0]
    const {container, unmount} = render(<ArticleComment comment={articleComment}/>)
    expect(container.outerHTML).toMatchSnapshot()
    expect(screen.getByText(articleComment.body)).toBeInTheDocument()
    expect(screen.getByText(articleComment.email)).toBeInTheDocument()
    unmount()
  })
})
