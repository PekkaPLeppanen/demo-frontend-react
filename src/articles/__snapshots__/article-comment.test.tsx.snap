// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ArticleComment /> render 1`] = `"<div><div class=\\"article-comment\\"><div class=\\"article-comment__email\\">Eliseo@gardner.biz</div><div class=\\"article-comment__body\\">laudantium enim quasi est quidem magnam voluptate ipsam eos tempora quo necessitatibus dolor quam autem quasi reiciendis et nam sapiente accusantium</div></div></div>"`;
